[Clang] [C++26] Expansion Statements (Part 8: Codegen)

Sirraide opened the pull request:
This implements codegen for expansion statements and adds some codegen tests.

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 8)

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 8)

Sirraide commented on clang/lib/CodeGen/CGStmt.cpp:
```diff
@@ -1511,6 +1512,44 @@ CodeGenFunction::EmitCXXForRangeStmt(const CXXForRangeStmt &S,
+  JumpDest ExpandExit = getJumpDestInCurrentScope("expand.end");
+  JumpDest ContinueDest;
+  for (auto [N, Inst] : enumerate(S.getInstantiations())) {
+    if (!HaveInsertPoint()) {
```
Yeah, that check doesn’t make sense

Sirraide commented on clang/lib/CodeGen/CGStmt.cpp:
```diff
@@ -1511,6 +1512,44 @@ CodeGenFunction::EmitCXXForRangeStmt(const CXXForRangeStmt &S,
+    else
+      ContinueDest = getJumpDestInCurrentScope("expand.next");
+
+    LexicalScope ExpansionScope(*this, S.getSourceRange());
```
> Have you given any thought to what debug info should look like? It seems like it might be a little confusing to just point at the line number without any context.

Hmm, I’m not really too familiar w/ how we typically handle debug info; it should more or less behave like a sequence of template instantiations; what exactly needs to be done to accomplish this here, I’m not sure

Sirraide commented on clang/test/CodeGenCXX/cxx2c-expansion-stmts-control-flow.cpp:
```diff
@@ -0,0 +1,429 @@
+// RUN: %clang_cc1 -std=c++2c -triple x86_64-unknown-linux-gnu -emit-llvm -o - %s | FileCheck %s
```
This seems to just work; I’ll add some tests

shafik commented on clang/lib/CodeGen/CGStmt.cpp:
```diff
@@ -1511,6 +1512,36 @@ CodeGenFunction::EmitCXXForRangeStmt(const CXXForRangeStmt &S,
 
+void CodeGenFunction::EmitCXXExpansionStmtInstantiation(
+    const CXXExpansionStmtInstantiation &S) {
+  // FIXME: For reasons beyond my understanding, two scopes are required to emit
```
@erichkeane Any ideas? I see some the CIRGEN openacc stuff has nested scopes too although not back to back like this.

erichkeane replied:
Oh wow, this IS odd... This absolutely shouldn't be necessary.  This really feels like we're missing a scope somewhere that this is working around, but I don't have a good idea of where/why.

Sirraide replied:
Yeah, I have no idea what’s happening here; there’s supposed to be a compound statement around the entire expansion statement, and that one _isn’t_ present in the AST, but one lexical scope ought to be enough for that... so I don’t know why it isn’t

efriedma-quic replied:
The double-scope thing came up in #199508.  It's related to the way pushLifetimeExtendedDestroy works: SD_Automatic is translated in CodeGen to mean "destroy this when the parent scope is popped".

I think the symptom is pointing to something different in this context, though: I'd guess template instantiation skips creating an ExprWithCleanups somewhere.

Sirraide replied:
> I'd guess template instantiation skips creating an ExprWithCleanups somewhere.

Oh that might be it; I’ll investigate that

Sirraide replied:
Looks like that was the problem: I was missing a call to `MaybeCreateExprWithCleanups()` in `TransformCXXExpansionStmtPattern()`; thanks! I’m not sure I how long it would have taken me to figure that out...

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 9: Control Flow) (#169688)

This disallows certain `switch`-`case` constructs and labels in
expansion statements.

GNU local labels are supported as an extension. My reasoning for
allowing this is that:
1. I’d argue it makes sense: the main reason for disallowing labels is
that you’d end up w/ more than one label w/ the same name in a single
function, and jumping to such a label from outside an expansion
statement becomes rather ill-defined as a result (which instantiation
does it jump to? what if the statement expands to nothing at all?). Both
of these issues don’t exist with local labels since they aren’t scoped
to the entire function and as a result can only be jumped to from inside
their respective instantiation.
2. Codegen and constant evaluation for this basically just works.
3. Disallowing this in the most straight-forward manner (e.g. by just
dropping the label) leads to rather poor QOI and a lot of diagnostics,
so even if we wanted to disallow local labels... we’d basically still
have to do everything I’m doing in this patch in addition to then
issuing a diagnostic for good QOI.
